crates/protocol/interop/src/graph.rs

```diff
@@ -87,57 +87,27 @@ where
     /// terminal case being all blocks reduced to deposits-only.
     ///
     /// [int-block-replacement]: https://specs.optimism.io/interop/derivation.html#replacing-invalid-blocks
-    pub async fn resolve(mut self) -> MessageGraphResult<(), P> {
+    pub async fn resolve(self) -> MessageGraphResult<(), P> {
         info!(
             target: "message_graph",
             "Checking the message graph for invalid messages"
         );
 
-        // Reduce the graph to remove all valid messages.
-        self.reduce().await?;
-
-        // Check if the graph is now empty. If not, there are invalid messages.
-        if !self.messages.is_empty() {
-            // Collect the chain IDs for all blocks containing invalid messages.
-            let mut bad_block_chain_ids =
-                self.messages.into_iter().map(|e| e.executing_chain_id).collect::<Vec<_>>();
-            bad_block_chain_ids.dedup();
-
-            warn!(
-                target: "message_graph",
-                bad_chain_ids = %bad_block_chain_ids
-                    .iter()
-                    .map(ToString::to_string)
-                    .collect::<Vec<_>>()
-                    .join(", "),
-                "Failed to reduce the message graph entirely",
-            );
-
-            // Return an error with the chain IDs of the blocks containing invalid messages.
-            return Err(MessageGraphError::InvalidMessages(bad_block_chain_ids));
-        }
-
-        Ok(())
-    }
-
-    /// Attempts to remove as many edges from the graph as possible by resolving the dependencies
-    /// of each message. If a message cannot be resolved, it is considered invalid. After this
-    /// function is called, any outstanding messages are invalid.
-    async fn reduce(&mut self) -> MessageGraphResult<(), P> {
         // Create a new vector to store invalid edges
-        let mut invalid_messages = Vec::with_capacity(self.messages.len());
+        let mut invalid_messages = HashMap::default();
 
-        // Prune all valid edges.
-        for message in core::mem::take(&mut self.messages) {
-            if let Err(e) = self.check_single_dependency(&message).await {
+        // Prune all valid messages, collecting errors for any chain whose block contains an invalid
+        // message. Errors are de-duplicated by chain ID in a map, since a single invalid
+        // message is cause for invalidating a block.
+        for message in self.messages.iter() {
+            if let Err(e) = self.check_single_dependency(message).await {
                 warn!(
                     target: "message_graph",
                     executing_chain_id =  message.executing_chain_id,
                     message_hash = ?message.inner.payloadHash,
-                    "Invalid ExecutingMessage found",
+                    "Invalid ExecutingMessage found: {e}",
```

Comment:
nit: instead of formatting inside the warn string we could put the error as a field like 
```suggestion
                    warn!(
                    target: "message_graph",
                    executing_chain_id =  message.executing_chain_id,
                    message_hash = ?message.inner.payloadHash,
                    err = e,
                    "Invalid ExecutingMessage found",
```